### ISSUE TITLE
Update Native Maps to use more recent C++ standards 

### ISSUE THREAD

**Issue by phrocker** [apache/accumulo#1386]
It may also help us to move away from raw pointers for internal objects where we can use RAII to better manage resources. Only top level objects need a java managed lifetime. everything else should use smart pointers. this alleviates any potential concern for memory leaks and also allow us to better manage object lifetimes. 

Further, I think we can better exploit move semantics and improve performance. As a final task in this, we should look at copy semantics where they are employed. It seems that memory is copied quite a bit. In some cases it is not avoidable, but it would behoove us to evaluate their usages.

**Comment by phrocker** [comment]
I think it is fair to stick with C++11 as that would give us the most bang for the buck and also alleviate one issue I found with the "Iterator" struct and reference management.

**Comment by keith-turner** [comment]
> I think it is fair to stick with C++11

I am so out of the loop on C++.  Why do you think C++11 is a good goal?  I am guessing that is whats currently available in recent Centos and Ubuntu releases.

**Comment by phrocker** [comment]
> > I think it is fair to stick with C++11
> 
> I am so out of the loop on C++. Why do you think C++11 is a good goal? I am guessing that is whats currently available in recent Centos and Ubuntu releases.

C++14 and C++17 are well supported on most distros, but there are some older compilers on u14/16 that only support certain portions of C++11. Realistically those are EOL, but there is nothing new  in C++14 or C++17 that I would use other than std::make_unique ( https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique ). 

TL;DR I don't want to impact the requirements for compilation in a minor release, so I'm going with the lowest common denominator that has the greatest ROI.

**Comment by ctubbsii** [comment]
@phrocker Is this something you're thinking about working on?

**Comment by phrocker** [comment]
@ctubbsii Yah I've started... since it's something I hope to use as a reference artifact in my c++ client (https://github.com/phrocker/sharkbite), I hope I'm not a liar and finish it in a reasonable amount of time.

**Comment by mjwall** [comment]
[What is RAII?](https://en.cppreference.com/w/cpp/language/raii)

**Comment by SlickNik** [comment]
@phrocker Would love to pair on this with you if possible! It's been a long while since I've brushed off my C++, and I've been itching to give some of the newer RAII/move semantics pieces of C++11 (and later) a shot. Would also be a good avenue for me to learn more about Accumulo's use of Native Maps.

**Comment by phrocker** [comment]
@SlickNik absolutely I'll get some thought together on breaking this up. Thanks!

**Comment by mjwall** [comment]
I will be watching for this too. Let me know if you need any help reviewing or whatever.

**Comment by phrocker** [comment]
@SlickNik Here is the current changeset https://github.com/phrocker/accumulo/compare/master...nativeupdates2x?expand=1

Note that I've brought in two dependencies. One is a lock free concurrent queue that will facilitate work down the road to improve concurrency of native maps in general but it also allows me to use a singleton pattern for memory delegation. I'm performing some benchmarks now, but in contrived tests I see improvements; however, the ultimate test will be with more realistic work loads. I need to clean the branch up and add some of my tests. I'd like to move to cmake so the two added third party dependencies can be downloaded; however, I may make that a follow on task.

**Comment by phrocker** [comment]
I'll do some cleanup tomorrow morning then submit a WIP PR so you all can take a look in a PR.

**Comment by SlickNik** [comment]
Thanks for pushing this up @phrocker! I've pulled down the changes from your branch and should have some time to go through them over the next day or two. I haven't yet looked at how this is all being built, so I may have some more questions as I do that - maybe the move to cmake is something that I can help with as well, as I look into this.

**Comment by ctubbsii** [comment]
If anybody has questions about how the native map builds currently work, I can probably answer those... although I think it's relatively simple: Essentially, we have a maven module that, instead of producing a jar at the package phase, produces a tarball containing a `Makefile` and our C++ sources (including those generated from `javac -h` earlier in the multi-module maven build). This tarball is included inside the `-bin.tar.gz` distribution file for users to extract and build for their platform. I think we provide a convenience script to unpack and call `make` for them (in case `tar xvf ... && cd ... && make` is too hard :smiley_cat:)

You can see this tarball by running `mvn clean package -DskipTests` and looking at `server/native/target/accumulo-native-*.tar.gz`

I would be interested in understanding the benefits of moving to `cmake`. As I understand it, most of the benefits come from the ability to reduce the usage of other autotools/configure stuffs, but we don't use any of that. We just have a *very* trivial `Makefile`.

The focus is on user-buildability, rather than providing pre-built native binaries. If we do switch to `cmake`, it seems like something that could happen as its own independent task, but I would want to maintain our current practice of making it user-buildable, rather than pre-building binaries.

**Comment by phrocker** [comment]
@ctubbsii I agree that the make file is relatively simple, but I'm afraid it's not about simplicity. 

In the case where you have external projects that you are linking and using ( in my branch I am ), there is an ExternalProject command which not only simplifies that but helps ensures that you don't link incorrectly. It also helps avoiding pollution of your source especially when you are only including it in binary form ( as we are ). 

Additionally, I saw some version specific variables embedded in the make file. This can go away because CMAKE takes all of that into account and automates the generation of your make file. 

Even more so we use -O3...frankly we should be allowing this to be 
overridden. There are other modes ( -Os if that were preferred in certain deployments ) that have practical applications. Is a rule of thumb we should probably use a lower optimization level (-O3 ) level with select flags specified. My hope is to prove this via google benchmark. Managing those flags across GCC versions would be simpler with CMAKE, because that is part of its job. 

The final point is that adding dependencies like google bench mark and google test would be far simpler with CMAKE for not only the reason above, but because we can manage those dependencies. I'm certainly not suggesting we take a stab at the cmake in this issue, but I included two additional dependencies and added them to a "thirdparty" directory That's not unreasonable, but those source files do not need to be included even by way of a git sub module. ExternalProject is preferred IMO. 

In the same vein as google benchmark I did some tests with jemalloc which showed dramatic improvement in certain use cases. Showing those benchmarks and allowing a static link into the SO would be ideal and much easier with CMAKE. All of this can be done with make, but why reinvent the wheel?

Your point of "we just have a very trivial Makefile" is spot on, but when you begin to change that, using prior work versus writing our own allocator for example, I think the argument for cmake begins to gain traction. Further, as you integrate things like google benchmark to prove, then keeping the Makefile is more an effort in adding work, complexity, and hidden bugs than incorporating something that has minimal impact. It would be stated that this would be totally transparent to the user and something hidden maven.

**Comment by ctubbsii** [comment]
Those are all very good reasons to use cmake. I'm not opposed. I just want to understand the benefits.

I do wonder if it's a good idea to add third party dependencies to code which is currently very small, and has very minimal dependencies. Even if it's a performance benefit... or a maintenance benefit (less code)... it may not be worth the trade off. It'd be nice if that could be done as an independent task.

As I see it, there are 3 things being discussed (and worked on) here:

1. Change the build system for the native code (make -> cmake and/or improvements to the hard-coded flags in the Makefile)
2. Modernize the code (using more recent C++ standards)
3. Reduce some of our custom code with third party libraries that do the job better

In my opinion, all three of these are possibly good improvements, but they should be decoupled from one another. We're talking about making all these changes at once, and I think they each have their own pros and cons, and they should be done completely independently. That will help facilitate code reviews, and discussions about pros/cons of specific change proposals.

**Comment by phrocker** [comment]
@ctubbsii to be fair I'm not proposing cmake at this time or for this task. I'm only doing task two(modernization) but the fallout is that item three is quite natural (avoiding custom code, where other code is maintained to avoid bugs with compilers, is cross platform, etc. ) I'm excluding Google benchmark at this time due to not wanting to handle dependencies in make.

**Comment by SlickNik** [comment]
@ctubbsii Totally agree that issue 1 is orthogonal to 2 and 3. There may be some coupling between 2 and 3 (in some cases where the more modern code may depend on 3rd party libs) - and @phrocker has more visibility to that than I do, so he is in a better position to talk about that.

Also agree that it's a good idea tackle a move to cmake, or any other such build improvements separately. That will make for easier separation of concerns during  discussions, reviews, etc.

**Comment by ctubbsii** [comment]
I think it should be a goal to minimize the coupling between 2 and 3, if it can be done, but understand that it may not be possible (or sensible).
